app/screens/settings/notification_mention/mention_settings.tsx

```diff
@@ -63,48 +49,119 @@ const getStyleSheet = makeStyleSheetFromTheme((theme) => {
         container: {
             paddingHorizontal: 8,
         },
+        input: {
+            color: theme.centerChannelColor,
+            height: 150,
+            paddingHorizontal: 15,
+            ...typography('Body', 100, 'Regular'),
+        },
+        containerStyle: {
+            marginTop: 30,
+        },
     };
 });
 
+const getMentionProps = (currentUser: UserModel) => {
+    const notifyProps = getNotificationProps(currentUser);
+    const mKeys = (notifyProps.mention_keys || '').split(',');
+
+    const usernameMentionIndex = mKeys.indexOf(currentUser.username);
+    if (usernameMentionIndex > -1) {
+        mKeys.splice(usernameMentionIndex, 1);
+    }
+
+    return {
+        mentionKeys: mKeys.join(','),
+        usernameMention: usernameMentionIndex > -1,
+    };
+};
+
 type MentionSectionProps = {
-    currentUser?: UserModel;
-    mentionKeys: string;
+    componentId: string;
+    currentUser: UserModel;
 }
-const MentionSettings = ({currentUser, mentionKeys}: MentionSectionProps) => {
-    const [{firstName, usernameMention, channel}, dispatch] = useReducer(reducer, INITIAL_STATE);
+const MentionSettings = ({componentId, currentUser}: MentionSectionProps) => {
+    const serverUrl = useServerUrl();
+
+    const notifyProps = currentUser.notifyProps as Partial<UserNotifyProps>;
+    const [tglFirstName, setTglFirstName] = useState(notifyProps.first_name === 'true');
+    const [tglChannel, setTglChannel] = useState(notifyProps.channel === 'true');
+
+    const mentionProps = useMemo(() => getMentionProps(currentUser), [currentUser.notifyProps]);
+    const [tglUserName, setTglUserName] = useState(mentionProps.usernameMention);
+    const [mentionKeys, setMentionKeys] = useState(mentionProps.mentionKeys);
+
     const theme = useTheme();
     const styles = getStyleSheet(theme);
     const intl = useIntl();
 
-    const toggleChannelMentions = () => {
-        dispatch({
-            type: UPDATE_MENTION_PREF,
-            data: {
-                channel: !channel,
-            },
-        });
-    };
-    const toggleUsernameMention = () => {
-        dispatch({
-            type: UPDATE_MENTION_PREF,
-            data: {
-                usernameMention: !usernameMention,
-            },
-        });
-    };
-    const toggleFirstNameMention = () => {
-        dispatch({
-            type: UPDATE_MENTION_PREF,
-            data: {
-                firstName: !firstName,
-            },
+    const saveButton = useMemo(() => {
+        return {
+            id: SAVE_MENTION_BUTTON_ID,
+            enabled: false,
+            showAsAction: 'always' as const,
+            testID: 'notification_settings.mentions.save.button',
+            color: theme.sidebarHeaderTextColor,
+            text: intl.formatMessage({id: 'settings.save', defaultMessage: 'Save'}),
+        };
+    }, [theme.sidebarHeaderTextColor]);
+
+    const canSave = useCallback(() => {
```

Comment:
I don't think this needs to be memoized

Reply:
The way this is used, it needs memoization. That being said, I think it makes more sense to handle this (the update of the button) with an effect, instead of a function. Any reason not to do it with an effect?

Reply:
corrected